### PR TITLE
Fix binding injector bug

### LINKED ISSUE #373: Major bug in service binding injector
I can't get through the walkthrough. Diving deep into the root cause, I've found that the k8s service injector uses this bit of code to add credential fields to a secret:

```
secret := &v1.Secret{
	ObjectMeta: v1.ObjectMeta{
		Name:      binding.Name,
		Namespace: binding.Spec.InstanceRef.Namespace,
	},
	Data: make(map[string][]byte),
}

// For each item in the cred just serialize its value into JSON and
// save it in the secret
for k, v := range *cred {
	data, err := json.Marshal(v)
	if err != nil {
		return fmt.Errorf("Unable to marshal credential value %q: %s",
			k, err)
	}
	secret.Data[k] = data
}
```

The problem with this is demonstrated [here in the go playground](https://play.golang.org/p/ntmiX8Z1ap)

The bytes resulting from marshaling a string _include_ leading and trailing double-quotes. These quotes are included in what is base64-encoded within the secret. The practical consequence is that while username and password `admin` / `foo` may have been provisioned, the consumer will end up using username and password `"admin"` / `"foo"` and authentication to the service instance will thus fail.

### PR DESCRIPTION
**Pull request by krancour**
Fixes #373

### CODE REVIEW

File: pkg/controller/injector/k8s_binding_injector.go [hunk 1]
```diff
@@ -64,15 +63,10 @@ func (b *k8sBindingInjector) Inject(binding *servicecatalog.Binding, cred *broke
...
+	// For each item in the cred, just convert it to the bytes for a string
+	// representation of the value
 	for k, v := range *cred {
-		data, err := json.Marshal(v)
-		if err != nil {
-			return fmt.Errorf("Unable to marshal credential value %q: %s",
-				k, err)
-		}
-		secret.Data[k] = data
+		secret.Data[k] = []byte(fmt.Sprintf("%v", v))
```
Comment (duglin) [inline pkg/controller/injector/k8s_binding_injector.go]: using %v scares me a little. If we're going to do this can we add, as part of this PR, a set of new testcases that verifies that all of the different types of JSON objects that might appear as values are round-tripped properly?

**Comment by duglin** [comment]
design SGTM, just think we need some serious testcases to make sure %v does what we want in all cases.

File: pkg/controller/injector/k8s_binding_injector.go (hunk 1 above)
Comment (krancour) [inline pkg/controller/injector/k8s_binding_injector.go]: @duglin can you articulate what sort of types you might be concerned about? A value that's anything other than a primitive doesn't make sense here and `%v` is going to work fine for all primitive types... and if you don't buy that, then it's golang that needs more test cases; not us.

Comment (duglin) [inline pkg/controller/injector/k8s_binding_injector.go]: What I mean is we should add tests for creds that come back like:
```
"user1": "password",
"user2": { "key" : "value" },
"user3": [ "value" ],
"user4": { "key" : { "key2":"value" }}
...
```
Just the various ways the json might appear.  In particular to make sure the nesting of types is printed correctly when using %v.  Regardless of whether any potential bug is in golang or not, we'd be the ones responsible for finding a work-around.

Comment (krancour) [inline pkg/controller/injector/k8s_binding_injector.go]: Ugh. Ya. I'd somehow talked myself into believing creds couldn't _possibly_ be arbitrarily nested JSON structures like that, but after double-checking the OSB spec, they can be. fml

**Comment by pmorie** [comment]
@duglin 

> design SGTM, just think we need some serious testcases to make sure %v does what we want in all cases.

@krancour and I talked in depth today about this and did some investigation in go playgrounds.  TLDR:

1.  We will handle primitives and unstructured JSON data
2.  We will have tests that demonstrate all happy-land and corner-cases

We also need to have a spec for how this binding responses are transformed for broker authors, so they will be able to predict how things will be transformed into k8s primitives.

**Comment by krancour** [comment]
@pmorie after further exploration of this, I want to amend the strategy we discussed.

As it turns out, strings, and strings alone are the corner case. _Everything_ else can be serialized using the same technique as is used today. This simplifies the logic considerably. It turns into simply, "if it's a string, deal with it differently."

But yes... I am bulking up the test cases here so that we test round-trips on this serialization for various primitive types, maps, and slices.

Standby for code drop.

**Comment by krancour** [comment]
I still need to document this better; i.e. add more comments in code to explain what's going on.

File: pkg/controller/injector/serializer_test.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	"reflect"
+	"testing"
+)
+
+// Tests in this file test this package's serialize() function by "round
+// tripping." Values are serialized using the serialize() function, then are
+// deserialized using the inverse of the process used by serialize(). Comparison
+// of the original value to the output of the round trip is used to assert the
+// correcness of the serialize() function.
+
```
Comment (pmorie) [inline pkg/controller/injector/serializer_test.go]: dead line

File: pkg/controller/injector/serializer_test.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+		}
+	}
+}
+
+func TestSerializeString(t *testing.T) {
+	strVals := []string{
+		"foo",
+		"bar",
+		"bat",
+		"bat",
```
Comment (pmorie) [inline pkg/controller/injector/serializer_test.go]: might want to put some unicode in here or fuzz some strings; also duplicate 'bat' case

**Comment by arschles** [comment]
@krancour LGTM - the docs look fine to me

File: pkg/controller/injector/serializer_test.go (hunk 3 above)
Comment (krancour) [inline pkg/controller/injector/serializer_test.go]: @pmorie I fixed the duplicated "bat". (I had intended for one to be "baz", but fat-fingered it.) I've also added some unicode-- "卧虎藏龙".

Fuzzing seems like it would be overkill, no?

**Comment by pmorie** [comment]
>  I would still like to see a table test with different serialized payloads that then checks the content of the created secret without relying on any of the encoding/decoding machinery.

I would be happy for this in a follow-up, if we can verify what's in here now makes the walkthrough usable again.

File: pkg/controller/injector/serializer_test.go (hunk 3 above)
Comment (arschles) [inline pkg/controller/injector/serializer_test.go]: >Fuzzing seems like it would be overkill, no?

FWIW I agree that it's overkill _for now_. Until we have a fully-functioning product (controller that talks to API server, complete working demo), I'd rather focus time on making that work than beefing up unit tests. Maybe make an issue for adding a fuzz test here?

Comment (krancour) [inline pkg/controller/injector/serializer_test.go]: Actually... overkill or not, fuzzing might be easy enough to add here that I don't mind doing so. Let me look.

**Comment by krancour** [comment]
@pmorie I added fuzzing to the serialization tests.

File: pkg/controller/injector/serializer_test.go (hunk 2 above)
Comment (krancour) [inline pkg/controller/injector/serializer_test.go]: I don't see this as a dead line. The comment is very general about _all_ of the tests that follow; not about any _one_ of them. I feel, therefore, that there _should_ be space separating it from what follows so it doesn't appear too specifically associated with the specific bit that follows and, by extension, less associated with the rest  of it.

**Comment by krancour** [comment]
> I would still like to see a table test with different serialized payloads that then checks the content of the created secret without relying on any of the encoding/decoding machinery.

That's more properly its own issue anyway, because what you're suggesting is well beyond the scope of this PR to begin with. i.e. It has no bearing on fixing the bug in the credential serialization logic. I'll rely on you to open that issue because I don't exactly see what's to be gained by it... so I think you can articulate the need better than I could.

File: pkg/controller/injector/serializer_test.go (hunk 2 above)
Comment (pmorie) [inline pkg/controller/injector/serializer_test.go]: Ah, you're right - I had misread the comment.
